### PR TITLE
Builder umessage

### PR DESCRIPTION
**Pull request by ruchirchauhan**
This PR implements builder for UMessage

### CODE REVIEW

File: src/datamodel/builder/UMessage.cpp
Comment (billpittman) [inline src/datamodel/builder/UMessage.cpp:180]: Same as before but for UPayload, I suggest UPayloadFormat_MIN & UPayloadFormat_MAX

File: src/datamodel/builder/UMessage.cpp
Comment (billpittman) [inline src/datamodel/builder/UMessage.cpp:148]: Was there a test for this throw ?  And the other domain_error throws ??  I looked, but might have missed it.

**Review by billpittman — CHANGES_REQUESTED** [review]
Overall looks good, some minor change requests and a few additional test requests (or clarifications).

File: src/datamodel/builder/UMessage.cpp
Comment (billpittman) [reply src/datamodel/builder/UMessage.cpp:148]: Hehehe - I guess I jumped the gun a bit, I just now noticed that this PR is a draft -- I probably should have waited until you were ready for reviewers -- Whoops.

Comment (ruchirchauhan) [reply src/datamodel/builder/UMessage.cpp:148]: Ohh. no worries, reviews are always welcome. In fact thank you for taking time. But yeah I am still working on unit tests.

File: src/datamodel/builder/UMessage.cpp
Comment (ruchirchauhan) [reply src/datamodel/builder/UMessage.cpp:180]: I agree.

Done

File: src/datamodel/builder/UMessage.cpp
Comment (billpittman) [inline src/datamodel/builder/UMessage.cpp:195]: Is this check backwards or am I confused.  If I set payload_format to UPAYLOAD_FORMAT_PROTOBUF, the code throws an error .... right?

Comment (ruchirchauhan) [reply src/datamodel/builder/UMessage.cpp]: Yes if the code has set any payloadFormat before calling build() (one with no parameters), this should throw error

**Review by billpittman — COMMENTED** [review]
WRT the tests in UMessageBuilderTest, I'm a little conflicted.  I like the tests you have, but I think there needs to be a few tests added to "verify" the creation of the resulting message.  Right now the tests verify that the build command completes without error and throws when expected, but it seems like there should be a few more tests that actually test what build() built.  However there's a significant overlap between those tests and the UMessageValidationTest that I'm currently working on.  Maybe we can pull some of the UMessageValidator tests into UMessageBuilder tests .... thoughts ?

**Comment by ruchirchauhan** [comment]
Since this build creates a message with empty payload and just attributes, there are not many tests that we need to do on it but it would make sense to add one more test , that tests the actual message been built for correct attributes.

File: src/datamodel/builder/UMessage.cpp
Comment (pranavishere2) [inline src/datamodel/builder/UMessage.cpp:221]: Please avoid using set_allocated method because it needs to allocate memory using new operator. Try using mutable instead. Here is the code:

UMessageBuilder::UMessageBuilder(v1::UMessageType msgType,
                v1::UUri&& source,
                v1::UUri&& sink,
                v1::UUID&& request_id) {

    attributes_.set_type(msgType);

    *attributes_.mutable_source() = std::move(source);

    if (sink.IsInitialized()) {
        *attributes_.mutable_sink() = std::move(sink);
    }

    attributes_.set_priority(v1::UPriority::UPRIORITY_CS1);

    if (request_id.IsInitialized()) {
        *attributes_.mutable_reqid() = std::move(request_id);
    }

    attributes_.set_payload_format(static_cast<v1::UPayloadFormat>(v1::UPayloadFormat_MIN - 1));
}

Comment (ruchirchauhan) [reply src/datamodel/builder/UMessage.cpp:221]: @gregmedd @billpittman, what do you think about this ?

Comment (billpittman) [reply src/datamodel/builder/UMessage.cpp:227]: No super strong opinion really.  Are the objects being copied into the function and then again when we set their values with set_allocated_**** ?  So two copies, then if so, I think it would be wise to use the mutable version.

**Comment by billpittman** [comment]
@ruchirchauhan For the sake of our maintainers, would you put a comment/reference in the description about how/why this PR should be merged after that-PR-link.

File: src/datamodel/builder/UMessage.cpp
Comment (gregmedd) [inline src/datamodel/builder/UMessage.cpp:223]: Use of `new` is forbidden. Use local variables or use a smart pointer (`std::make_unique` or `std::make_shared`) if dynamic memory allocation is required.

File: src/datamodel/builder/UMessage.cpp
Comment (gregmedd) [reply src/datamodel/builder/UMessage.cpp:227]: I agree with @pranavishere2. Manual memory management (e.g. via `new`) should _never_ be done without _very good reason_.

For the record, we are unlikely to ever have a good reason in up-cpp.

File: include/up-cpp/datamodel/builder/UMessage.h
Comment (gregmedd) [inline include/up-cpp/datamodel/builder/UMessage.h:245]: Use `std::optional` to clearly convey whether a parameter was set or not. Default construction of message objects may not be the right behavior and is hard to check.

File: src/datamodel/builder/UMessage.cpp
Comment (gregmedd) [inline src/datamodel/builder/UMessage.cpp:197]: This is where storing the intermediate values in a UAttributes poses a challenge. It would be better if there was a `std::optional<v1::UPayloadFormat> format_` in the class. Then it's easy to tell if it has ever been set (`if (format_)...`).

File: src/datamodel/builder/UMessage.cpp
Comment (gregmedd) [inline src/datamodel/builder/UMessage.cpp:200]: Something more like _tried to build with no payload when a payload format has been set_ would probably be more clear.

File: src/datamodel/builder/UMessage.cpp
Comment (gregmedd) [inline src/datamodel/builder/UMessage.cpp:213]: This will need to populate the message payload

File: test/coverage/datamodel/UMessageBuilderTest.cpp
Comment (gregmedd) [inline test/coverage/datamodel/UMessageBuilderTest.cpp:32]: Why not just
```suggestion
        source_.set_authority_name("10.0.0.1");
```

File: src/datamodel/builder/UMessage.cpp
Comment (gregmedd) [inline src/datamodel/builder/UMessage.cpp:204]: The message ID needs to be generated for every message.

File: test/coverage/datamodel/UMessageBuilderTest.cpp
Comment (gregmedd) [inline test/coverage/datamodel/UMessageBuilderTest.cpp:29]: This is called for each test case, but the values do not change. Is this intentional?

**Review by gregmedd — COMMENTED** [review]
I have reviewed the main code, but not the test code yet. There are some changes. My main concerns are around the use of `new`.

File: src/datamodel/builder/UMessage.cpp
Comment (ruchirchauhan) [reply src/datamodel/builder/UMessage.cpp:227]: Done

File: src/datamodel/builder/UMessage.cpp
Comment (ruchirchauhan) [reply src/datamodel/builder/UMessage.cpp:223]: Done

File: include/up-cpp/datamodel/builder/UMessage.h
Comment (ruchirchauhan) [reply include/up-cpp/datamodel/builder/UMessage.h:245]: Done, however I still have to provide default std::nullopt to it. So essentially, am still doing similar thing only using a different style..

File: src/datamodel/builder/UMessage.cpp
Comment (ruchirchauhan) [reply src/datamodel/builder/UMessage.cpp:204]: Yeah, I realized that. done

File: test/coverage/datamodel/UMessageBuilderTest.cpp
Comment (ruchirchauhan) [reply test/coverage/datamodel/UMessageBuilderTest.cpp:32]: done

File: test/coverage/datamodel/UMessageBuilderTest.cpp
Comment (ruchirchauhan) [reply test/coverage/datamodel/UMessageBuilderTest.cpp:29]: What would be a better way to do this kind of setup ?

File: src/datamodel/builder/UMessage.cpp
Comment (ruchirchauhan) [reply src/datamodel/builder/UMessage.cpp:200]: Done

File: src/datamodel/builder/UMessage.cpp
Comment (ruchirchauhan) [reply src/datamodel/builder/UMessage.cpp:213]: I was blocked on this because payload was not implemented. Now that Pranav is working on it, I will sync with him and update this code.

File: src/datamodel/builder/UMessage.cpp
Comment (ruchirchauhan) [reply src/datamodel/builder/UMessage.cpp:197]: Done

File: test/coverage/datamodel/UMessageBuilderTest.cpp
Comment (gregmedd) [reply test/coverage/datamodel/UMessageBuilderTest.cpp:29]: If the values are completely constant, then they can be `constexpr` variables. If they need to be set up once for the whole test run, then then can be set in `SetUpTestSuite()`. If using variables set at runtime by one of the test methods, then care should be taken that tests cannot alter the values (e.g. provide accessors or make sure they are const).

Comment (ruchirchauhan) [reply test/coverage/datamodel/UMessageBuilderTest.cpp:29]: Done

**Review by gregmedd — CHANGES_REQUESTED** [review]
A few changes are still required, and you're blocked on . Additionally, there appears to be a CI issue. I would recommend rebasing to main to make sure you have the latest code.

While rebasing, I recommend using interactive rebase (`rebase -i`) to squash your commits.

File: src/datamodel/builder/UMessage.cpp
Comment (ruchirchauhan) [reply src/datamodel/builder/UMessage.cpp]: done

File: src/datamodel/builder/UMessage.cpp
Comment (ruchirchauhan) [reply src/datamodel/builder/UMessage.cpp]: fixed

File: include/up-cpp/datamodel/builder/UMessage.h
Comment (ruchirchauhan) [reply include/up-cpp/datamodel/builder/UMessage.h]: done

File: src/datamodel/builder/UMessage.cpp
Comment (ruchirchauhan) [reply src/datamodel/builder/UMessage.cpp]: Done, but this required removal of const qualifier at the end of the function because now thiis function needs to modify the state of the object it is called on.

File: src/datamodel/builder/UMessage.cpp
Comment (ruchirchauhan) [reply src/datamodel/builder/UMessage.cpp]: fixed

File: src/datamodel/builder/UMessage.cpp
Comment (ruchirchauhan) [reply src/datamodel/builder/UMessage.cpp]: Done

**Comment by gregmedd** [comment]
@ruchirchauhan - it looks like #161 is almost ready to merge. Once that's in, you should be able to rebase and we can (hopefully) wrap up this PR too.

File: src/datamodel/builder/UMessage.cpp
Comment (gregmedd) [inline src/datamodel/builder/UMessage.cpp:220]:
```suggestion
	*message.mutable_payload() = std::move(payloadData);
	message.mutable_attributes()->set_payload_format(payloadFormat);
```

File: src/datamodel/builder/UMessage.cpp
Comment (gregmedd) [inline src/datamodel/builder/UMessage.cpp:213]: Switch to move semantics for extracting the payload. Also, #161 is renaming `build()` and `move()` to `buildCopy()` and `buildMove()` for clarity:

```suggestion
	auto [payloadData, payloadFormat] = std::move(payload).buildMove();
```

File: src/datamodel/builder/UMessage.cpp
Comment (gregmedd) [reply src/datamodel/builder/UMessage.cpp]: It shouldn't need to modify the state of the object. It can set the value in the message's copy of attributes after the copy is made. So instead of `*attributes_.mutable_id()...` you would do `*(message.mutable_attributes()->mutable_id())...`

**Review by gregmedd — COMMENTED** [review]
Aside from a few small remaining changes and integration with the final PayloadBuilder from #161, this looks good to me. Commit history is good. Once the changes are done and #161 is in, I will approve and merge.

**Comment by gregmedd** [comment]
@ruchirchauhan - PR #161 has been merged, so you should be able to use the PayloadBuilder now. Please rebase onto main to get the latest changes.

File: src/datamodel/builder/UMessage.cpp
Comment (ruchirchauhan) [reply src/datamodel/builder/UMessage.cpp:213]: done

File: src/datamodel/builder/UMessage.cpp
Comment (ruchirchauhan) [reply src/datamodel/builder/UMessage.cpp]: Makes sense, done

File: include/up-cpp/datamodel/builder/UMessage.h
Comment (gregmedd) [inline include/up-cpp/datamodel/builder/UMessage.h:220]:
```suggestion
	[[nodiscard]] v1::UMessage build() const;
```

File: include/up-cpp/datamodel/builder/UMessage.h
Comment (gregmedd) [inline include/up-cpp/datamodel/builder/UMessage.h:257]:
```suggestion
	UuidBuilder uuidBuilder_;
```

File: src/datamodel/builder/UMessage.cpp
Comment (gregmedd) [inline src/datamodel/builder/UMessage.cpp:214]: Order is reversed here, and also the UUID builder's build can be used directly since you have an instance of it already:
```suggestion
	*message.mutable_attributes() = attributes_;
	*(message.mutable_attributes()->mutable_id()) =
	    uuidBuilder_.build();
```

File: src/datamodel/builder/UMessage.cpp
Comment (gregmedd) [inline src/datamodel/builder/UMessage.cpp:204]:
```suggestion
	*message.mutable_attributes() = attributes_;
	*(message.mutable_attributes()->mutable_id()) =
	    uuidBuilder_.build();
```

File: src/datamodel/builder/UMessage.cpp
Comment (gregmedd) [inline src/datamodel/builder/UMessage.cpp:230]: `uuidBuilder_` is not initialized
```suggestion
UMessageBuilder::UMessageBuilder(v1::UMessageType msgType, v1::UUri&& source,
                                 std::optional<v1::UUri>&& sink,
                                 std::optional<v1::UUID>&& request_id) 
                                 : uuidBuilder_(UuidBuilder::getBuilder()) {
```

**Review by gregmedd — COMMENTED** [review]
@ruchirchauhan - really sorry about this, but I have one more batch of small changes that need to be made. They are all in the form of diffs so they should be easy to go through and apply. Thanks!

**Review by billpittman — APPROVED** [review]
Thanks for adding the extra tests.  If this PR is still gated by other PRs, please put the ones it is gated by in the comment to make it clear that it should wait to be merged.

I've been out for a week, so theoretically those PRs that were gating this one have all been submitted, if so, then there's nothing to do.  I'm approving the PR, because all my concerns have been met -- THANKS!

**Review by gregmedd — APPROVED** [review]
LGTM

**Pull request merged**
